commit 681f4820ab7339b0ecc4cf656ae1e0adac6ed8f6
Author: Jeremy Huddleston <jeremyhu@apple.com>
Date:   Mon Oct 24 16:21:28 2011 -0700

    apple: Implement applegl_unbind_context
    
    glXMakeCurrent(dpy, None, NULL) would not correctly unbind the context
    causing subsequent GLX requests to fail in peculiar ways
    
    http://xquartz.macosforge.org/trac/ticket/514
    
    Signed-off-by: Jeremy Huddleston <jeremyhu@apple.com>
    (cherry picked from commit 5c44c1348ea13f51a1616968daa7034bb48e42b1)

diff --git a/src/glx/applegl_glx.c b/src/glx/applegl_glx.c
index 8766c88..9be12b0 100644
--- a/src/glx/applegl_glx.c
+++ b/src/glx/applegl_glx.c
@@ -69,6 +69,24 @@ applegl_bind_context(struct glx_context *gc, struct glx_context *old,
 static void
 applegl_unbind_context(struct glx_context *gc, struct glx_context *new)
 {
+   Display *dpy;
+   bool error;
+
+   /* If we don't have a context, then we have nothing to unbind */
+   if (!gc)
+      return;
+
+   /* If we have a new context, keep this one around and remove it during bind. */
+   if (new)
+      return;
+
+   dpy = gc->psc->dpy;
+
+   error = apple_glx_make_current_context(dpy,
+					  (gc != &dummyContext) ? gc->driContext : NULL,
+					  NULL, None);
+
+   apple_glx_diagnostic("%s: error %s\n", __func__, error ? "YES" : "NO");
 }
 
 static void
